### PR TITLE
Channel/Bucket interface refactor

### PR DESCRIPTION
**Pull request by beaucollins**
The `Channel` object is too nosy and overrides instance methods on the `Bucket` which leads to indirect code paths.

These changes remove the knowledge of a `Bucket` instance from the `Channel`. It is now up to the `Bucket` to use the correct `Channel` APIs and listen for the correct `Channel` events.

From a 3rd party's perspective the api has not changed.

## Bucket Promise API

The `Bucket` interface now provides a `Promise` based interface. All methods that take a callback also return a `Promise`:

```js
// Still works
bucket.get( 'my-object', callback );

// Now you can
bucket.get( 'my-object' ).then( success, failure );

// And async/await if you want to get rid of all the indents
const object = await bucket.get( 'my-object' );
```

## Additional Cleanup

- Stops listening to its own events. Event emit/listener pairs aren't necessary for an object that both generates and listens for those events.
- Now that a bucket calls methods directly on a `Channel` we can get rid of `bucketEvents`
- Bucket/Channel API is more clearly defined through `.prototype.` methods on `Channel`
- JSDoc added to methods called by the `Bucket` (and some additional ones as well).
- Snuck in the `node-uuid` to `uuid` dependency update

### CODE REVIEW

File: src/simperium/bucket.js [hunk 1]
```diff
@@ -1,60 +1,253 @@
...
+ * @returns {Object} store api methods that use Promises instead of callbacks
+ */
+const promiseAPI = store => ( {
+	get: ( id ) =>
+		callbackAsPromise( store.get.bind( store, id ) ),
+	update: ( id, object, isIndexing ) =>
+		callbackAsPromise( store.update.bind( store, id, object, isIndexing ) ),
+	remove: ( id ) =>
+		callbackAsPromise( store.remove.bind( store, id ) ),
+	find: ( query ) =>
```
Comment (roundhill) [inline src/simperium/bucket.js]: How do you feel about removing the parens for one parameter? `find: query =>`

File: src/simperium/bucket.js [hunk 2]
```diff
@@ -1,60 +1,253 @@
...
+ * @param {Function} task - function that expects a single callback argument
+ * @returns {Promise} callback wrapped in a promise interface
+ */
+const callbackAsPromise = ( task ) => new Promise( ( resolve, reject ) => {
+	task( ( error, result ) => {
+		if ( error ) {
+			reject( error );
+			return;
+		}
+		resolve( result );
```
Comment (roundhill) [inline src/simperium/bucket.js]: What if you one lined this block:
```
error ? reject( error ) : resolve( result );
```

**Comment by beaucollins** [comment]
This should be backwards compatible.

**Comment by roundhill** [comment]
> This should be backwards compatible.

Sorry, I meant maybe we don't need to add the compatibility code? Move forward, blaze a new trail and all that.

**Comment by beaucollins** [comment]
There's two pieces to consider for backwards compatibility:

- The `Bucket` API keeping callbacks along with promises
- The `BucketStore` API that uses callbacks that this PR now changes the API for internally to a Promise based one.

Simplenote currently uses both of these.

I think a separate PR to make those breaking changes makes sense.

**Comment by beaucollins** [comment]
@roundhill I just gave this  a quick run through with Simplenote and it appeared to be working. I didn't exercise a lot of the app.

**Comment by roundhill** [comment]
> @roundhill I just gave this a quick run through with Simplenote and it appeared to be working. I didn't exercise a lot of the app.

ok, I will give it a test in Simplenote as well.

**Comment by dmsnell** [comment]
wow I wish I had looked here before spending way too long trying to understand the circular references in the channel and bucket messaging!

@roundhill there's always old versions of the npm package that aren't going away if people can't update yet. we can major-bump-it.

**Pull request merged**